namespaces/resource-namespace.md

```diff
@@ -0,0 +1,150 @@
+
+### Data Structure (MSON Element)
+
+Data structure definition using MSON namespace elements.
+
+## Namespace Elements
+
+### Resource (Element)
+
+The Resource representation with its available transitions and attributes.
+
+#### Properties
+
+- `element`: resource (string, fixed)
+- `attributes` (object)
+    - `href` (Href) - URI Template of this resource.
+    - `hrefVariables` (Href Variables) - Definition of URI Template variables used in the `href` property.
+
+- `content` (array)
+    - (Transition) - State transition available for this resource
+    - (Data Structure) - Data structure associated with the resource
+
+#### Example
+
+```json
+{
+  "element": "resource",
+  "meta": {
+    "title": "Question",
+    "description": "A Question object has the following attributes."
+  },
+  "attributes": {
+    "href": "/questions/{question_id}",
+    "hrefVariables": {
+      "element": "object",
+      "content": [
+        {
+          "element": "member",
+          "content": {
+            "key": {
+              "element": "string",
+              "content": "question_id"
+            },
+            "value": {
+              "element": "number",
+              "content": "1"
+            }
+          }
+        }
+      ]
+    }
+  },
+  "content": [
+    {
+      "element": "object",
+      "meta": {
+        "title": "Question"
+      },
+      "content": [
+        {
+          "element": "member",
+          "content": {
+            "key": {
+              "element": "string",
+              "content": "question"
+            },
+            "value": {
+              "element": "string",
+              "content": "Favourite programming language?"
+            }
+          }
+        }
+      ]
+    }
+  ]
+}
+```
+
+### Transition (Element)
+
+A transition is an available progression from one state to another state. Exercising a transition initiates HTTP transaction.
+
+#### Properties
+
+- `element`: transition (string, fixed)
+- `attributes` (object, required)
+    - `relation` - (string) - The relation of this transition from the current resource.
+- `content` (object, required)
+    - `href` (Href, required) - The URI template for this transition.
+    - `parameters` (Href Variables) - A hash of input parameters for this transition.
+    - `attributes` (Data Structure) - Input attributes for this transition.
+    - `transactions` (array[Transaction]) - Array of transaction examples.
+
+#### Example
+
+```json
+{
+    "element": "transition",
+    "attributes": {
+      "relation": "update"
+    },
+    "content": {
+        "href": "https://polls.apiblueprint.org/questions/1"
+    }
+}
+```
+
+## Protocol Specific
+
+### HTTP Transaction (Element)
+
+Example of an HTTP Transaction. Note this structure can be extended in a future to support other protocols.
+
+#### Properties
+
+- `element`: transaction (string, fixed)
+- `attributes`
+    - `protocol` (fixed)
+        - http (string)
+- `content` (array[HTTP Message])
```

Comment:
Right now this looks like RFC 7230, which makes me think we'll get something like this:

```json
        [
          "resource",
          {
            "title": "Question"
          },
          {},
          [
            [
              "transition",
              {
                "title": "View a Question Detail"
              },
              {
                "href": "/questions/{question_id}"
              },
              [
                [
                  "httpTransaction",
                  {},
                  {},
                  [
                    "GET /questions/1234 HTTP/1.1\r\nHost: example.com\r\n",
                    "200 OK\r\nContent-Type: application/json\r\n\r\n{\"question\": \"test\"}\r\n"
                  ]
                ]
              ]
            ]
          ]
        ]
```

It seems strange to include the RFC 7230 document inside of our existing structure. I like the simplicity of the format, but we lose semantic information that the parser/library now needs to use heuristics for, e.g. is this a request or a response? We are also putting the burden of being able to parse RFC 7230 on the parser/library author. I almost think it would be better to have a canonical HTTP message, preferably just a simple hash but it could also be refracted.

```json
"GET /questions/1234 HTTP/1.1\r\nHost: example.com\r\n"
# Becomes
{
  "method": "GET",
  "path": "/questions/1234",
  "headers": {
    "Host": "example.com"
  },
  "body": null
}
```